src/makerspace/views.py

```diff
@@ -67,3 +70,48 @@ class EquipmentDeleteView(PermissionRequiredMixin, PreventGetRequestsMixin, Dele
     permission_required = ('makerspace.delete_equipment',)
     model = Equipment
     success_url = reverse_lazy('admin_equipment_list')
+
+
+class StatisticsView(TemplateView):
+    template_name = 'makerspace/statistics.html'
+
+    q_overnight = Sum(F('start_time__hour') - F('end_time__hour'))
+    get_time_in_hours = 3600000000  # number the reservation length needs to be divided by to get the length in hours
+    reservations = Reservation.objects.prefetch_related(Prefetch('machine', queryset=Machine.objects.prefetch_related(
+        'machine_type'))).filter(special=False).annotate(q_overnight=q_overnight)
+
+    def get_machines_with_reservations(self, machine_type):
+        filter_reservations = Q(reservations__special=False) & Q(reservations__event__isnull=True)
+        return Machine.objects.prefetch_related('machine_type', 'reservations').filter(Q(machine_type__name__icontains=machine_type) & filter_reservations).order_by('machine_type')
+
+    def get_time_distribution(self):
+        overnight = self.reservations.filter(q_overnight__gte=0)  # reservations overnight has to be counted differently
+        not_overnight = self.reservations.exclude(q_overnight__gte=0)
+        time = {}
+        for r in range(0, 24):
+            time[r] = not_overnight.filter(Q(start_time__hour__lte=r) & Q(end_time__hour__gte=(r - 1))).count()
+            time[r] += overnight.filter(Q(start_time__hour__lte=r) | Q(end_time__hour__gte=(r - 1))).count()
+            if self.reservations.count() != 0:
+                time[r] = time[r]/self.reservations.count()
+        return dict(time.items())
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        span_of_printer_reservations = list(self.get_machines_with_reservations("3D-printere").
+                                            annotate(len=Sum((F('reservations__end_time') - F('reservations__start_time')) / self.get_time_in_hours, output_field=IntegerField())).values('id', 'name', 'len'))
+        number_of_printer_reservations = list(self.get_machines_with_reservations("3D-printere").annotate(number_of_reservations=Count('reservations')).values('id', 'name', 'number_of_reservations'))
+        longest_printer_reservations = list(self.get_machines_with_reservations("3D-printere").
+                                            annotate(len=Sum((F('reservations__end_time') - F('reservations__start_time')) / self.get_time_in_hours, output_field=IntegerField())).
+                                            order_by('-len').values('id', 'name', 'reservations', 'len'))[:3]
```

Comment:
Oh wait, after testing some of the code locally, it seems like `longest_printer_reservations` does indeed contain what the name suggests, and the annotated `len` attribute's value is, appropriately, the duration of each reservation - in contrast to [the renaming suggestion I made in a comment above](https://github.com/MAKENTNU/web/pull/669#discussion_r1403761054) 😅 The code makes this pretty confusing, since each row/element of the list does _not_ represent each machine - the opposite of what you'd expect, and what the lists above and below contain. _(I'm explaining this to avoid any potential misunderstandings:)_ Instead, due to the way `values()` works, when providing the name of a many-to-many field or a reverse many-to-one manager (`'reservations'` in this case), it will return one row/element for each combination the multivalued fields provided (one row per reservation, in this case), as explained in the warning at the bottom of [the `values()` section in the docs](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#values). Under [the `values_list()` section](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#values-list), this is mentioned:
> **`values()`** and **`values_list()`** are both intended as optimizations for a specific use case: retrieving a subset of data without the overhead of creating a model instance. This metaphor falls apart when dealing with many-to-many and other multivalued relations (such as the one-to-many relation of a reverse foreign key) because the “one row, one object” assumption doesn’t hold.

So, I think code relying on such an obscure fact of the `values()` method should be avoided (or at the very least explicitly pointed out and explained in comments). I think I found a fairly easy replacement, however, which also makes the query slightly less complex by going through the reservation table instead of the machine table: 🙂(This contains some elements from other parts of the code and from my other suggestions, as you can see :))
```python
longest_printer_reservations = (
    Reservation.objects.filter(event=None, special=False, machine__machine_type__pk=1)
    .annotate(duration=Sum(F('end_time') - F('start_time')))
    .order_by('-duration')
    .select_related('machine')
    [:3]
)
longest_printer_reservations_list = []
for reservation in longest_printer_reservations:
    duration: timedelta = reservation.duration
    longest_printer_reservations_list.append({
        'machine_name': reservation.machine.name,
        'duration_hours': duration.total_seconds() // (60 * 60),
    })
```
`longest_printer_reservations_list` can then be inserted into the context variables 🙂 Any thoughts?